File: pastprograms/USACO/Silver/2016O-P2.py
fin = open("diamond.in", 'r')
fout = open("diamond.out", 'w')

N, K = map(int, fin.readline().split(" "))
sizes = sorted([int(fin.readline()) for i in range(N)])
ends = []
right = 0
for i in range(N):
    while(right < N):
        if(sizes[right] - sizes[i] > K): break
        right += 1
    ends.append(right - i)
taken = []
maxto = [0]
for e in ends[::-1]: maxto.insert(0, max(maxto[0], e))
best = 0
for i in range(N):
    current = ends[i]
    best = max(best, current + maxto[min(N, i + current)])
fout.write(str(best))

File: pastprograms/USACO/Silver/2016O-P2.cpp
#define pb push_back
#define mt make_tuple
#include <bits/stdc++.h>
using namespace std;

int N, M, a;
vector<int> nums;

int main() {
    ifstream fin;
    fin.open("diamond.in");
    ofstream fout;
    fout.open("diamond.out");
    fin >> N >> M;
    for (int i = 0; i < N; i ++) {
        fin >> a;
        nums.pb(a);
    }
    sort(nums.begin(), nums.end());
    vector<int> sizes, msizes;
    for (int i = 0; i < N; i ++) {
        int ni = nums[i];
        int os = lower_bound(nums.begin(), nums.end(), ni + M + 1) - nums.begin();
        sizes.pb(os - i);
    }
    sizes.pb(0);
    int rs = 0;
    for (int i = N; i >= 0; i --) {
        rs = max(rs, sizes[i]);
        msizes.pb(rs);
    }
    int best = 0;
    for (int i = 0; i < N; i ++) {
        int si = sizes[i];
        int startpos = i + si;
        best = max(best, si + msizes[(N - startpos)]);
    }
    fout << best;
}